Support include

lucifer1004 opened the pull request:
Target issue: #605 

- [x] AST
- [x] ASR and others (via `--preprocess-include` option and the preprocessor)

@certik Please review.

certik commented:
Excellent! Yes, we need this.

However, I am a little worried that it might not be possible to implement in full generality at the AST level, that include must be implemented at the prescanning or preprocessing level. I need to think about this and look at some test cases to see if the AST approach can support all of it. If it cannot, then we need to do this earlier in the pipeline.

Either way, thanks for working on this, we'll get this resolved soon. I'll get back to you soon.

certik commented:
The include feature is indeed hard to format correctly. It's a similar issue like the preprocessor, it's hard to format it correctly by going into AST and back.

If `include` can always be included in AST, then we should use that approach. The problem is that you might get syntax errors sometimes. In which case how would we decide if we should include it in AST versus do it already in the prescanning phase?

certik commented:
I am getting this warning in `build0.sh`:
```
preprocessor.re:546:78: warning: tag 't2' has 2nd degree of nondeterminism [-Wnondeterministic-tags]
```

certik commented:
Here is one problem:
```console
$ cat a.f90 
program test
include "b.f90"
real(dp) :: x
x = 5
print *, x
end program
$ cat b.f90 
integer, parameter :: dp = kind(0.d0)
```
gets reformatted as:
```console
$ lfortran fmt a.f90                
program test
real(dp) :: x
include 'b.f90'
x = 5
print *, x
end program test
```
and will fail, since `dp` is not declared "soon enough". The issue is that "include" can appear anywhere, not just as a statement, but also declaration, and at the global / module level.

I think in general include should be discouraged. For legacy codes, we should support this feature in its entirety, which I think should be using the preprocessor.

certik commented on src/bin/lfortran.cpp:
```diff
@@ -1496,7 +1496,7 @@ int main(int argc, char *argv[])
         app.add_flag("--implicit-typing", compiler_options.implicit_typing, "Allow implicit typing");
         app.add_flag("--allow-implicit-interface", compiler_options.implicit_interface, "Allow implicit interface");
-
+        app.add_flag("--preprocess-include", compiler_options.preprocess_include, "Preprocess Fortran includes");
```
I think this should be the default.

certik commented:
Here is a simple example that shows that in principle `include` can't be included as an AST node:
```console
$ cat a.f90 
program test
integer :: i
i = 5
include "b.f90"
$ cat b.f90 
print *, i
end program
```
Which works in the preprocessor:
```console
$ lfortran --preprocess-include a.f90 --show-ast --indent
(TranslationUnit 
    [(Program 
        test 
        () 
        [] 
        [] 
        [(Declaration 
            (AttrType 
                TypeInteger 
                [] 
                () 
                None
            ) 
            [] 
            [(i 
            [] 
            [] 
            () 
            None 
            ())] 
            ()
        )] 
        [(= 
            0 
            i 
            5 
            ()
        )
        (Print 
            0 
            () 
            [i] 
            ()
        )] 
        []
    )]
)
```
But fails in this PR's default mode:
```console
$ lfortran  a.f90 --show-ast --indent
syntax error: End of file is unexpected here
 --> a.f90:5:1
  |
5 | 
  | ^ 


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
```

certik commented:
To move forward, why don't you just keep the preprocessing parts in this PR, make it the default, let's add some tests, polish and merge.

Then in a separate PR, if you are still interested in pursuing the idea of an AST include node, we can explore the idea and the associated limitations of that approach.

certik commented:
Do you want to implement the preprocessor include in a separate PR? I need it for SciPy, so it would save me time if you want to finish it.

lucifer1004 commented:
A problem here is that the current method for preprocessing include is copied from C preprocessing which claims to be only working on Linux.

lucifer1004 commented:
@certik Do we have tests involving includes? Seems that all tests have passed, except for a parser issue. Since I am not  familiar with `re2c`, can you give me a hand to fix the warning that failed the CI?

lucifer1004 commented:
So the include statement is still needed (#777)? I have removed it from this PR, and I also removed the `--preprocess-include` option since I thought include preprocessing was going to be done by default. Seems I misunderstood you. Should I revert the latest commit in this PR, and then move the AST part into a separate PR?

certik commented:
This is good. I didn't realize that you added it into the preprocessor, which indeed is not run by default. I think it should not be run by default. I think we have to add it to the prescanner, which takes a source code, and generates another source code (in memory). We have to run the prescanner anyway, as the Fortran tokenizer can't easily handle some of stuff that the prescanner takes care of. And so we should handle the `include` in there.

lucifer1004 commented:
Actually, I added a new function `preprocess_include()` into the preprocessor, instead of modifying the `run()` function. The `run()` function just works as before.

lucifer1004 commented:
Since the current implementation of `preprocess_include` relies on `re2c`, more effort is needed to move it into the pre-scanner.

lucifer1004 commented:
I just realized there would be some repeated calls of `fix_continuation()` and `preprocess_include()`.

- We need to first run `fix_continuation` (because some include statements might be in multiple lines)
- We then need to run `preprocess_include`
- The included file might as well need `fix_continuation`
- And it might also contain nested includes.
- ...

lucifer1004 commented:
It is a bit difficult to make this preprocessing pass compatible with the C preprocessing pass since we have to refactor the LocationManager to get the correct mapping.

certik commented:
Yes, it's a bit complicated. Our location manager doesn't handle multiple files yet (I think), so we just mark everything imported via the include to have location span of the include itself, which is good enough to get started, later we'll fix it.

I was going to run this as part of `fix_continuation`, but you are right about the possibility of split include with the `&`. I need to check if it is allowed. If so, then we need to run this after `fix_continuation`. One option is to run this in the tokenizer itself, but then we would have to recursively switch to the new tokenizer for the include, which seems complex. 

There is a performance issue too: we do not want to run these passes unless it is needed. That is why they are hidden behind the if statement. In free-form for example if you don't split tokens with the `&` approach, you do not need to run prescanner, and if there are no preprocessor directives, you do not need to run the preprocessor. So the idea would be that these might be off by default, and we get very fast parsing. If a given file uses the preprocessor or the fancy (very rare) token splitting using `&`, then one has to run the additional passes by enabling them with a compiler options, in exchange for slower compilation.

certik commented:
I would do something like: https://github.com/lfortran/lfortran/pull/847

lucifer1004 commented:
Refactored based on #847

lucifer1004 commented:
To summarize, I did the following things:

- I added `IncludeStmt` to AST. When `prescan = false`, include statements will be recognized as AST nodes, while there is no `visit_Include()` in ASR so if you do `--show-asr` or something further, an error will occur claiming that `visit_Include()` not implemented.
- I implemented the include detection in fixed/free form based on #847, with minor modifications to make it actually work
- I implemented include with `std::filesystem` or `std::experimental::filesystem`, depending on C++ & OS versions, with the help of some conditions added to `lfortran/utils.h` copied from an online resource. I also modified the C preprocessor's include handler to also use this to enable multi-platform support.

lucifer1004 commented:
There seems to be something wrong with the filesystem usage.

lucifer1004 commented:
It is strange that the test failed on Ubuntu, while it passed on MacOS.

certik commented:
Very good tests.

Can you please implement your changes in the [parser.cpp](https://github.com/lfortran/lfortran/pull/762/files#diff-55c8b282f6fc0b64d1d867ce5e5d61980166b55d608f19cab44077d221227e3d) without reformatting the file? It's hard for me to see what the actual changes are and if they are correct.

Since the include never makes it to AST in this PR, I would remove the other changes and put them into a separate PR. I would do the bare minimum to get things working in this PR in the prescanner level to get things working, we'll polish it, add tests (you already have them) and merge. Then we can improve upon it in various ways.

certik commented:
Regarding the errors, it fails on my machine (macOS M1) too, here is a commit that updates tests:

https://github.com/lfortran/lfortran/commit/7156ca22376bacad2d4f0c1e0aadd4a5a8dcbf68

From it you can see the errors are:
```console
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
[2m  File "[0m[1m[35m$DIR/src/bin/lfortran.cpp[39m[0m[2m", line 1667[0m
    return emit_ast(arg_file, compiler_options);
[2m  File "[0m[1m[35m$DIR/src/bin/lfortran.cpp[39m[0m[2m", line 420[0m
    LFortran::Result<std::string> r = fe.get_ast(input, lm, diagnostics);
[2m  File "[0m[1m[35m$DIR/src/lfortran/fortran_evaluator.cpp[39m[0m[2m", line 163[0m
    Result<AST::TranslationUnit_t*> ast = get_ast2(code, lm,
[2m  File "[0m[1m[35m$DIR/src/lfortran/fortran_evaluator.cpp[39m[0m[2m", line 190[0m
    tmp = fix_continuation(*code, lm, compiler_options.fixed_form);
[2m  File "[0m[1m[35m$DIR/src/lfortran/parser/parser.cpp[39m[0m[2m", line 436[0m
    LFORTRAN_ASSERT(std::string(pos, pos + 7) == "include");
AssertFailed: std::string(pos, pos + 7) == "include"
```
and
```
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
[2m  File "[0m[1m[35m$DIR/src/bin/lfortran.cpp[39m[0m[2m", line 1667[0m
    return emit_ast(arg_file, compiler_options);
[2m  File "[0m[1m[35m$DIR/src/bin/lfortran.cpp[39m[0m[2m", line 420[0m
    LFortran::Result<std::string> r = fe.get_ast(input, lm, diagnostics);
[2m  File "[0m[1m[35m$DIR/src/lfortran/fortran_evaluator.cpp[39m[0m[2m", line 163[0m
    Result<AST::TranslationUnit_t*> ast = get_ast2(code, lm,
[2m  File "[0m[1m[35m$DIR/src/lfortran/fortran_evaluator.cpp[39m[0m[2m", line 190[0m
    tmp = fix_continuation(*code, lm, compiler_options.fixed_form);
[2m  File "[0m[1m[35m$DIR/src/lfortran/parser/parser.cpp[39m[0m[2m", line 512[0m
    LFORTRAN_ASSERT(check_newlines(s, lm.in_newlines))
AssertFailed: check_newlines(s, lm.in_newlines)
```

certik commented:
I marked it as Draft. Once it is ready, just click on "Ready for review".

lucifer1004 commented:
The format of the rest part of parser.cpp has been changed back. I still could not reproduce the test errors, both the one in CI and the one you provided. (They are two different tests, the failed CI test is the integration test, and the one failed on your computer is the test for fixed form)

certik commented:
I pushed in the updated reference tests. As you can see, the only ones that fail are the ones added in this PR. We just have to fix it.

certik commented:
I see, there is a difference in the outputs on macOS and linux. I've seen this before, there is some randomness in this PR then. Either some undefined behavior, or possibly some differences, although typically this happens between Windows / Linux, not macOS/Linux.

certik commented on src/lfortran/parser/parser.cpp:
```diff
@@ -359,6 +398,17 @@ std::string fix_continuation(const std::string &s, LocationManager &lm,
+                case LineType::Include: {
+                    while (pos < s.size() && s[pos] == ' ')
+                        pos++;
+                    LFORTRAN_ASSERT(std::string(pos, pos + 7) == "include");
```
This line somehow fails. I can help debug it tomorrow.

certik replied:
I went over the code, but can't see anything obvious. I think the `pos` is somehow wrong, possibly with nested includes.

lucifer1004 replied:
Maybe or maybe not. I really need to be able to reproduce the errors you and the CI met before any further investigations.

certik commented:
@lucifer1004 I would try to do something along the lines of https://github.com/lfortran/lfortran/pull/855.

certik commented:
@lucifer1004 I implemented the core of the feature in #859, using your code from this PR that I fixed to work, I put you as co-author on the relevant commits. What is missing is the support in AST. I thought about it, and I think for most (but not all) usages of include, we should be able to parse it to AST and then parse the include file itself to AST, and include at the AST level; in which case we do not need to run the prescanner at all. So we can have both.

If you are interested in that feature, you can submit it as a new PR.

lucifer1004 commented:
This PR has been separated into #859 and #860.
